modules/lang-expression/src/main/java/org/elasticsearch/script/expression/ExpressionSearchScript.java

```diff
@@ -89,14 +97,29 @@ public LeafSearchScript getLeafSearchScript(final LeafReaderContext leaf) throws
             @Override
             public void setDocument(int d) {
                 docid = d;
+                try {
+                    values.advanceExact(d);
+                } catch (IOException e) {
+                    throw new IllegalStateException("Can't advance to doc using " + compiledScript, e);
+                }
             }
 
             @Override
             public void setScorer(Scorer s) {
                 scorer = s;
                 try {
                     // We have a new binding for the scorer so we need to reset the values
-                    values = source.getValues(Collections.singletonMap("scorer", scorer), leaf);
+                    values = source.getValues(leaf, new DoubleValues() {
+                        @Override
+                        public double doubleValue() throws IOException {
+                            return scorer.score();
+                        }
+
+                        @Override
+                        public boolean advanceExact(int doc) throws IOException {
+                            return true;
+                        }
+                    });
```

Comment:
No we can't ;).
There is an assert in the scorer created in this function that checks that the scorer's docid is set when `advanceExact` is called. Though the canned scorer that we use set the docid when `score` is called which makes it incompatible with this helper.